smt/surrogate_models/krg_based.py

```diff
@@ -1082,59 +1094,69 @@ def grad_minus_reduced_likelihood_function(log10t):
 
                     for i in range(len(self.noise0)):
+                        v = i + len(self.theta0)
+                        v2 = deepcopy(v)
                         noise_bounds = np.log10(noise_bounds)
-                        constraints.append(
-                            lambda log10t: log10t[i + len(self.theta0)]
-                            - noise_bounds[0]
-                        )
-                        constraints.append(
-                            lambda log10t: noise_bounds[1]
-                            - log10t[i + len(self.theta0)]
-                        )
+                        constraints.append(lambda log10t: log10t[v2] - noise_bounds[0])
+                        constraints.append(lambda log10t: noise_bounds[1] - log10t[v2])
                         bounds_hyp.append(noise_bounds)
+                theta_limits = np.repeat(
+                    np.log10([theta_bounds]), repeats=len(theta0), axis=0
+                )
+                sampling = LHS(
+                    xlimits=theta_limits, criterion="maximin", random_state=41
+                )
+                theta_lhs_loops = sampling(self.options["n_start"])
+                theta_all_loops = np.vstack((theta0, theta0_rand, theta_lhs_loops))
                 try:
-
                     if self.options["hyper_opt"] == "Cobyla":
-                        optimal_theta_res = optimize.minimize(
-                            minus_reduced_likelihood_function,
-                            theta0,
-                            constraints=[
-                                {"fun": con, "type": "ineq"} for con in constraints
-                            ],
-                            method="COBYLA",
-                            options={"rhobeg": _rhobeg, "tol": 1e-4, "maxiter": limit},
-                        )
-
-                        optimal_theta_res_2 = optimal_theta_res
+                        for theta0_loop in theta_all_loops:
+                            optimal_theta_res_loop = optimize.minimize(
+                                minus_reduced_likelihood_function,
+                                theta0_loop,
+                                constraints=[
+                                    {"fun": con, "type": "ineq"} for con in constraints
+                                ],
+                                method="COBYLA",
+                                options={
+                                    "rhobeg": _rhobeg,
+                                    "tol": 1e-4,
+                                    "maxiter": limit,
+                                },
+                            )
+                            try:
+                                if (
+                                    optimal_theta_res_loop["fun"]
+                                    < optimal_theta_res["fun"]
+                                ):
+                                    optimal_theta_res = optimal_theta_res_loop
+                            except:
```

Comment:
I suppose you test for an exception because `optimal_theta_res` is not defined? For me it would be better to initialize it with something like:
```python
optimal_theta_res = { "fun": float("inf") }
```
But I do not know if it fits with the rest of the algorithm

Reply:
I thought it would be worse. I will do this, then. We definitely need a cleaner code. 
It was a pain to debug. There is 2 intricated loop for kriging and 4  for MFKPLSK...